### PR TITLE
Document JSON I/O error classes

### PR DESCRIPTION
**Pull request by izgzhen**
Issue https://github.com/ethereum/solidity/issues/2186

### CODE REVIEW

**Comment by izgzhen** [comment]
@axic Still WIP, but you may comment the first few examples at any time

**Comment by axic** [comment]
That seems to be a good location for it. You could add a comment in the output example next to `errorType` to "see below for complete list of types".

**Comment by izgzhen** [comment]
Finally grabbed sometime to finish the rest error names that I can find :)

File: docs/using-the-compiler.rst [hunk 1]
```diff
@@ -273,3 +274,16 @@ Output Description
...
+
+Error types
+~~~~~~~~~~~
+
+1. ``JSONError``: JSON input doesn't conform to the required format, e.g. input is not a JSON object, the language is not supported etc.
+2. ``IOError``: Input/Output errors, e.g. invalid URLs
+3. ``InternalCompilerError``: internal compiler error, e.g. compilation failed, JSON logic error, JSON runtime error etc.
+4. ``Exception``: Unknown exception during compilation
+5. ``UnimplementedFeatureError``: some feature is not implemented
+6. ``Exception``: Exception during compilation
```
Comment (axic) [inline docs/using-the-compiler.rst]: This is there twice

File: docs/using-the-compiler.rst [hunk 2]
```diff
@@ -273,3 +274,16 @@ Output Description
...
       }
     }
+
+
+Error types
+~~~~~~~~~~~
+
+1. ``JSONError``: JSON input doesn't conform to the required format, e.g. input is not a JSON object, the language is not supported etc.
+2. ``IOError``: Input/Output errors, e.g. invalid URLs
+3. ``InternalCompilerError``: internal compiler error, e.g. compilation failed, JSON logic error, JSON runtime error etc.
```
Comment (axic) [inline docs/using-the-compiler.rst]: JSON errors are not reported as part of this.

**Comment by axic** [comment]
These are missing:
- DeclarationError
- DocstringParsingError
- ParserError
- SyntaxError
- TypeError

File: docs/using-the-compiler.rst (hunk 2 above)
Comment (izgzhen) [reply docs/using-the-compiler.rst]: the information comes from code:

```
./libsolidity/interface/StandardCompiler.cpp:457:		return formatFatalError("InternalCompilerError", string("JSON logic exception: ") + _exception.what());
./libsolidity/interface/StandardCompiler.cpp:461:		return formatFatalError("InternalCompilerError", string("JSON runtime exception: ") + _exception.what());
```

Comment (axic) [reply docs/using-the-compiler.rst]: Ah ok, those messages are a bit misleading. It still means that the compiler code is broken, not that the supplied data is broken.

All it should say is something like "internal bug in the compiler (should be reported as an issue)".

Comment (izgzhen) [reply docs/using-the-compiler.rst]: I see, thanks a lot

**Review by pirapira — DISMISSED** [review]
Please fix
```
/home/travis/build/ethereum/solidity/docs/using-the-compiler.rst:290: WARNING: Enumerated list ends without a blank line; unexpected unindent.
```
in the Travis build.

**Comment by izgzhen** [comment]
updated, thanks!

**Comment by axic** [comment]
Added warnings and updated the descriptions.

**Review by axic — APPROVED** [review]

**Comment by axic** [comment]
The test failure was only a truffle issue. Merging this as the documentation test passed.

**Pull request merged**